Parsing improvements: no reactor, add parsing options, error severity options

auduchinok opened the pull request:
Parsing without queuing to the reactor may help with delays in getting possible breakpoint ranges, navigation bars in VS. We've been using it in Rider for some time already and it has greatly improved parsing and some editor features performance. It can also be used to implement typing assistance based on AST like code auto-formatting.

This PR introduces two new options types (parsing and error severity) that may be useful for FCS users. Parsing options have a reasonable default instance that only needs updating source files list. The default errors severity options may be used as it is.
Parsing options can be created from command line args so it is easy to replace FSharpProjectOptions in many scenarios.

A questionable change: the removal of DependencyFiles from FSharpParseFileResults. The only place it is used in is now-obsolete FSharpSource. If this field is needed somewhere else, I'll revert it. The copy of this list is still stored in type check results.

Currently this PR has changes to FCS related code only (fcs/FSharp.Compiler.Service.sln) and doesn't contain needed changes for VFT projects, so expect the VFT compilation and tests to fail. I'd like to add needed changes but would like this code to be reviewed and probably to have some assistance.

Fixes #3107 and https://github.com/fsharp/FSharp.Compiler.Service/issues/774

msftclas commented:
@auduchinok,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

realvictorprm commented:
Hey @auduchinok. Many many thanks for sharing this with the community!

vasily-kirichenko commented:
```
7>E:\github\visualfsharp\src\fsharp\vs\service.fs(1521,19): error FS1182: The value 'conditionalCompilationDefines' is unused
7>E:\github\visualfsharp\src\fsharp\vs\service.fs(1525,19): error FS1182: The value 'lexResourceManager' is unused
```

realvictorprm commented:
Can we (or more specifically me) help you fixing it or do you prefer to keep it clean @auduchinok?

auduchinok commented:
@vasily-kirichenko Thanks, accidentally copied these bits.
@realvictorprm Help with fixing VFT projects/tests would be appreciated.

saul commented:
Using @vasily-kirichenko's branch with the FSharp.Editor fixes and I'm getting syntax highlighting immediately in Project.fs - previously this would take an age to appear (if ever). This is a fantastic improvement to responsiveness.

auduchinok commented:
I merged @vasily-kirichenko's branch so it's now easier to try it in VS. However, it looks like parsing options are created wrongly in some places in these fixes, we'll look into it tomorrow.

cartermp commented:
Setting breakpoints deep in TypeChecker.fs, e.g.

https://github.com/Microsoft/visualfsharp/blob/f52362ad50e51f897d2b624adecc8c15cab56fbe/src/fsharp/TypeChecker.fs#L13286

Certainly feels faster to me.  I haven't been able the popup in #3404 to hit yet.  Nice!

@abelbraaksma, if you have the time, could you pull down this branch and see how it fares on your solution?

dsyme commented:
Some build errors: 
```
  D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\FSharpSource.fs(372,16): error FS0039: The field, constructor or member 'ParseFileInProject' is not defined. Maybe you want one of the following:   ParseAndCheckFileInProject   ParseFile   ParseAndCheckProject   CheckFileInProject [D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\FSharp.LanguageService.fsproj]
  D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\BackgroundRequests.fs(140,89): error FS0001: This expression was expected to have type    'FSharpParsingOptions'    but here has type    'FSharpProjectOptions' [D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\FSharp.LanguageService.fsproj]
  D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\BackgroundRequests.fs(152,55): error FS0039: The field, constructor or member 'ParseFileInProject' is not defined. Maybe you want one of the following:   ParseAndCheckFileInProject   ParseFile   ParseAndCheckProject   CheckFileInProject [D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\FSharp.LanguageService.fsproj]
  D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\BackgroundRequests.fs(159,63): error FS0039: The field, constructor or member 'ParseFileInProject' is not defined. Maybe you want one of the following:   ParseAndCheckFileInProject   ParseFile   ParseAndCheckProject   CheckFileInProject [D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\FSharp.LanguageService.fsproj]
  D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\BackgroundRequests.fs(187,58): error FS0039: The field, constructor or member 'ParseFileInProject' is not defined. Maybe you want one of the following:   ParseAndCheckFileInProject   ParseFile   ParseAndCheckProject   CheckFileInProject [D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\FSharp.LanguageService.fsproj]
  D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\BackgroundRequests.fs(207,85): error FS0039: The field, constructor or member 'DependencyFiles' is not defined. [D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\src\FSharp.LanguageService\FSharp.LanguageService.fsproj]
```

dsyme commented on src/fsharp/CompileOptions.fsi:
```diff
@@ -69,6 +69,9 @@ val GetCoreFscCompilerOptions     : TcConfigBuilder -> CompilerOptionBlock list
 val GetCoreFsiCompilerOptions     : TcConfigBuilder -> CompilerOptionBlock list
 val GetCoreServiceCompilerOptions : TcConfigBuilder -> CompilerOptionBlock list
 
+// Apply args to TcConfigBuilder and return new list of source files
```
Use ``///`` comments please

dsyme commented on src/fsharp/CompileOptions.fsi:
```diff
@@ -69,6 +69,9 @@ val GetCoreFscCompilerOptions     : TcConfigBuilder -> CompilerOptionBlock list
 val GetCoreServiceCompilerOptions : TcConfigBuilder -> CompilerOptionBlock list
 
+// Apply args to TcConfigBuilder and return new list of source files
+val ApplyCommandLineArgs: TcConfigBuilder * string list * string list -> string list
```
We don't yet enforce this, and don't do it often enough in the compiler, but please add argument names to the signature, e.g. 

```fsharp
val ApplyCommandLineArgs: x: TcConfigBuilder * y: string list * z: string list -> string list
```

dsyme commented on src/fsharp/vs/IncrementalBuild.fs:
```diff
@@ -1537,7 +1532,6 @@ type IncrementalBuilder(tcGlobals,frameworkTcImports, nonFrameworkAssemblyInputs
     member __.FileChecked = fileChecked.Publish
     member __.ProjectChecked = projectChecked.Publish
     member __.ImportedCcusInvalidated = importsInvalidated.Publish
-    member __.AllDependenciesDeprecated = allDependencies 
```
Is there a specific reason to remove this as part of this PR?  Just wondering, thanks

auduchinok replied:
No, not really, I've reverted it.

dsyme commented on src/fsharp/vs/service.fs:
```diff
@@ -1358,7 +1359,47 @@ type TypeCheckInfo
+    member x.LastFileName =
+        Debug.Assert(not (Array.isEmpty x.SourceFiles), "Parsing options don't contain any file")
+        Array.last x.SourceFiles
+    static member Default =
```
Please put a blank line between each ``member`` definition, thanks

dsyme commented on src/fsharp/vs/service.fs:
```diff
@@ -1372,153 +1413,120 @@ module internal Parser =
-         
+
         // This function gets called whenever an error happens during parsing or checking
-        let diagnosticSink sev (exn:PhasedDiagnostic) = 
```
Try to avoid the whitespace changes in a PR.  If you'd like to remove all trailing whitespaces, then please submit a single PR tht does it globally across all source files under ``src`` and ``vsintegration\src`` - that would be great - thanks!

auduchinok replied:
I would be easier to do it in a few parts, I think, rather than a single PR.
Should I revert these formatting changes in this PR?

dsyme replied:
@auduchinok it's not a big deal for this one - you can leave these changes in - it was just a comment for future PRs

dsyme commented on src/fsharp/vs/service.fs:
```diff
@@ -2289,10 +2286,7 @@ type BackgroundCompiler(legacyReferenceResolver, projectCacheSize, keepAssemblyC
-    let parseFileInProjectCache = 
-        MruCache<ParseCacheLockToken, _, _>(parseFileInProjectCacheSize, 
-            areSame=AreSameForParsing3,
-            areSimilar=AreSubsumable3)
```
Is it really ok to drop the ``areSimilar`` function here?

dsyme replied:
@auduchinok Could you comment on the removal of ``areSimilar`` here?  Is it needed? Perhaps now ``same`` and ``similar`` are the same?

auduchinok replied:
I think I shouldn't have removed it completely. I've reverted and changed it to check file name. Please tell if it's a wrong change.

dsyme replied:
That looks right

dsyme commented on src/fsharp/vs/service.fsi:
```diff
@@ -515,7 +528,9 @@ type internal FSharpChecker =
     member GetProjectOptionsFromCommandLineArgs : projectFileName: string * argv: string[] * ?loadedTimeStamp: DateTime * ?extraProjectInfo: obj -> FSharpProjectOptions
-           
+
+    member GetParsingOptionsFromCommandLineArgs: string list -> FSharpParsingOptions * FSharpErrorInfo list
```
Please add ``///`` comments and ``<param>`` entries, thanks!

dsyme commented on src/fsharp/vs/service.fsi:
```diff
@@ -385,21 +401,18 @@ type internal FSharpChecker =
-    /// <param name="userOpName">An optional string used for tracing compiler operations associated with this request.</param>
-    member ParseFileInProject : filename: string * source: string * options: FSharpProjectOptions * ?userOpName: string -> Async<FSharpParseFileResults>
+    /// <param name="options">Parsing options for the project or script.</param>
+    member ParseFile: filename: string * source: string * options: FSharpParsingOptions -> Async<FSharpParseFileResults>
```
Please add a ``ParseFileInProject`` for compatibility which extracts the parsing options from the project options.

auduchinok replied:
Should I mark it as deprecated?

auduchinok replied:
Should it have the same signature and include userOpName?

dsyme replied:
@auduchinok I think don't mark it as deprecated yet - we would need to update all docs.  Add an item on https://github.com/fsharp/FSharp.Compiler.Service/issues to track this

Yes, please use the same signature , and yes include userOpName - we should include that for ``ParseFile`` even if we don't use it as we should pass in causality-tracking information for any operation that consumes CPU thread or Reactor-thread resources

auduchinok commented:
@dsyme I've fixed most issues and reverted parsing APIs so existing code will fallback to the new APIs with creating parsing options.
Should I trace userOpName as well or looking at it in debug would be sufficient?
Thank you.

dsyme commented:
> Should I trace userOpName as well or looking at it in debug would be sufficient?

I suppose it would be best to add a single tracing line

dsyme commented:
@auduchinok There are a lot of failures, some listed here https://gist.github.com/dsyme/f75dde5e1bbc3f249566f55638766797

I think these are (mostly) to do with the removal of the added "new line" - tests are sensitive to that.  You should probably do that change in a separate PR.

dsyme commented on src/fsharp/vs/service.fs:
```diff
@@ -1372,153 +1418,120 @@ module internal Parser =
-          let errHandler = new ErrorHandler(reportErrors, mainInputFileName, tcConfig, source)
-
-          // Adding this new-line character at the end of the source seems odd but is required for some unit tests
-          let source = if source.Length = 0 || not (source.[source.Length - 1] = '\n') then source + "\n" else source 
```
This is the line that needs to be re-instated, or at least dealt with in a separate PR.  What failures did you find this caused?

Thanks

auduchinok replied:
It was causing access to non-existent lines in R# documents, which led to runtime exceptions.
It is possible to make a workaround for this hack but it clearly looks like a bug when a parser parses not a source you passed it but some different string.

auduchinok replied:
I've reverted it and filed a separate issue.

dsyme replied:
@auduchinok Yes, I agree. I've added the context I know to the linked bug.  On the bright side it could be worse - up until last year the same routine added **three** new line characters! :)

dsyme commented:
Here are the current "Errors and Failures" from the part 1 test log:
```
1) Error : Tests.LanguageService.TimeStamp.UsingMSBuild.Regression.NoContainedString.Timestamps.Bug3368a
System.Exception : Error seen containing "project1"
   at Tests.LanguageService.TimeStamp.UsingMSBuild.AssertNoErrorSeenContaining(OpenProject project, String text) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 44
   at Tests.LanguageService.TimeStamp.UsingMSBuild.Regression.NoContainedString.Timestamps.Bug3368a() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 119

2) Failed : Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ProjectReferenceAssemblyChange
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ProjectReferenceAssemblyChange() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 316

3) Failed : Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeAbsolute
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeAbsolute() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 208

4) Failed : Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeRelative
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeRelative() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 262

5) Error : Tests.LanguageService.TimeStamp.UsingProjectSystem.Regression.NoContainedString.Timestamps.Bug3368a
System.Exception : Error seen containing "project1"
   at Tests.LanguageService.TimeStamp.UsingMSBuild.AssertNoErrorSeenContaining(OpenProject project, String text) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 44
   at Tests.LanguageService.TimeStamp.UsingMSBuild.Regression.NoContainedString.Timestamps.Bug3368a() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 119

6) Failed : Tests.LanguageService.TimeStamp.UsingProjectSystem.Timestamps.ProjectReferenceAssemblyChange
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ProjectReferenceAssemblyChange() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 316

7) Failed : Tests.LanguageService.TimeStamp.UsingProjectSystem.Timestamps.ReferenceAssemblyChangeAbsolute
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeAbsolute() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 208

8) Failed : Tests.LanguageService.TimeStamp.UsingProjectSystem.Timestamps.ReferenceAssemblyChangeRelative
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeRelative() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 262
```
Some of the problems may be related to ``areSimilar``, I will take a look

dsyme commented on src/fsharp/vs/service.fs:
```diff
@@ -2352,49 +2357,23 @@ type BackgroundCompiler(legacyReferenceResolver, projectCacheSize, keepAssemblyC
             parseCacheLock.AcquireLock (fun ltok -> 
                 parseAndCheckFileInProjectCachePossiblyStale.Set(ltok, (filename,options),(parseResults,typedResults,fileVersion))  
-                parseAndCheckFileInProjectCache.Set(ltok, (filename,source,options),(parseResults,typedResults,fileVersion,priorTimeStamp))
-                parseFileInProjectCache.Set(ltok, (filename,source,options),parseResults))
```
It's possible some of the tests are failing because parseFileInProjectCache is not being populated at this point.  Is it possible to populate it?  I guess that would involve calling ``GetParsingOptionsFromProjectOptions`` which is costly, defeating the purpose of the cache

BTW note that ``parseFileInProjectCache`` should now be renamed to ``parseFileCache`` please

auduchinok replied:
I think it would be a good idea to store parsing options like it is currently done for project options and (probably?) pass it here as well.

dsyme commented on src/fsharp/vs/service.fs:
```diff
@@ -2289,10 +2297,7 @@ type BackgroundCompiler(legacyReferenceResolver, projectCacheSize, keepAssemblyC
-        MruCache<ParseCacheLockToken, _, _>(parseFileInProjectCacheSize, 
-            areSame=AreSameForParsing3,
-            areSimilar=AreSubsumable3)
+    let parseFileInProjectCache = MruCache<_,_,_>(parseFileInProjectCacheSize, areSimilar = AreSimilarForParsing, areSame = AreSameForParsing)
```
Please rename to ``parseFileCache``

Also please be explicit that this can be used from any caller thread, i.e. ``MruCache<AnyCallerThreadToken,_,_>`` like ``braceMatchCacheMru``

That said, I would feel a happier if we took a lock when accessing this cache as in the previous code - I don't think taking the lock on cache access will harm performance at all - and it might be better if the PR wasn't changing this cache from locked to lock free. I believe it's valid to do so (if not we have a bug in the braceMatchMruCache) - I'm just trying to minimize the size of the change in order to help you to get the tests to pass clean.

dsyme commented on src/fsharp/vs/service.fs:
```diff
@@ -2789,7 +2755,7 @@ type BackgroundCompiler(legacyReferenceResolver, projectCacheSize, keepAssemblyC
                 parseAndCheckFileInProjectCachePossiblyStale.Clear ltok
                 parseAndCheckFileInProjectCache.Clear ltok
-                parseFileInProjectCache.Clear ltok)
+                parseFileInProjectCache.Clear (AssumeAnyCallerThreadWithoutEvidence()))
```
As mentioned above, I think it's better for parseFileInProjectCache to still need a lock token

dsyme commented on src/fsharp/vs/service.fs:
```diff
@@ -2854,22 +2817,37 @@ type FSharpChecker(legacyReferenceResolver, projectCacheSize, keepAssemblyConten
+
+    member ic.ParseFileInProject(filename, source, options, ?userOpName: string) =
+        let userOpName = defaultArg userOpName "Unknown"
+        ic.ParseFile(filename, source, ic.GetParsingOptionsFromProjectOptions(options), userOpName)
```
Now I see the implementation of ``ParseFileInProject`` I'm concerned it's a bit costly, involving the full reparse of the command line arguments, setting up a TcConfigBuilder and so on.   Hmmm.... I suppose we should just deprecate this and adjust the Visual F# Tools to use ``ParseFile``

I will aim to take a look at that tomorrow.

dsyme commented on src/fsharp/vs/service.fsi:
```diff
@@ -515,7 +553,18 @@ type internal FSharpChecker =
+    /// <para>Get the FSharpParsingOptions implied by a set of command line arguments.</para>
+    /// </summary>
+    ///
+    /// <param name="projectFileName">Used to differentiate between projects and for the base directory of the project.</param>
```
These ``param`` entries are wrong for this method

forki commented:
ping

dsyme commented:
@forki  We need the tests to pass.  I can take a look at a few of the issues, but really the  change mentioned in this comment copied below should be done in a separate PR:

> I think these are (mostly) to do with the removal of the added "new line" - tests are sensitive to that. You should probably do that change in a separate PR.

auduchinok commented:
@dsyme I've already reverted it so it must be something else.
Currently I'm looking into what changing `IsSimilar` could break.

forki commented:
Tests.Service.ProjectAnalysisTests.Test Project21 whole project errors
Project21 error: <<<This expression was expected to have type
    'int'    
but here has type
    'string'    >>>

auduchinok commented:
@forki it is an output of the test, not the failure.
https://github.com/Microsoft/visualfsharp/blob/34e669cc05138d3e77b64302b20807717388d33f/tests/service/ProjectAnalysisTests.fs#L3009-L3014
An example of a failure is absence of completions in Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ProjectReferenceAssemblyChange.
The part2 of test suite was aborted due to overtime.

vasily-kirichenko commented:
@dotnet-bot test Windows_NT Release_ci_part1 Build please

vasily-kirichenko commented:
@dotnet-bot test Windows_NT Release_ci_part2 Build please

dsyme commented:
Getting closer... current "Errors and Failures" from part 1:
```
1) Error : Tests.LanguageService.TimeStamp.UsingMSBuild.Regression.NoContainedString.Timestamps.Bug3368a
System.Exception : Error seen containing "project1"
   at Tests.LanguageService.TimeStamp.UsingMSBuild.AssertNoErrorSeenContaining(OpenProject project, String text) in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 44
   at Tests.LanguageService.TimeStamp.UsingMSBuild.Regression.NoContainedString.Timestamps.Bug3368a() in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 119

2) Failed : Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ProjectReferenceAssemblyChange
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ProjectReferenceAssemblyChange() in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 316

3) Failed : Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeAbsolute
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeAbsolute() in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 208

4) Failed : Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeRelative
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeRelative() in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 262

5) Error : Tests.LanguageService.TimeStamp.UsingProjectSystem.Regression.NoContainedString.Timestamps.Bug3368a
System.Exception : Error seen containing "project1"
   at Tests.LanguageService.TimeStamp.UsingMSBuild.AssertNoErrorSeenContaining(OpenProject project, String text) in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 44
   at Tests.LanguageService.TimeStamp.UsingMSBuild.Regression.NoContainedString.Timestamps.Bug3368a() in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 119

6) Failed : Tests.LanguageService.TimeStamp.UsingProjectSystem.Timestamps.ProjectReferenceAssemblyChange
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ProjectReferenceAssemblyChange() in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 316

7) Failed : Tests.LanguageService.TimeStamp.UsingProjectSystem.Timestamps.ReferenceAssemblyChangeAbsolute
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeAbsolute() in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 208

8) Failed : Tests.LanguageService.TimeStamp.UsingProjectSystem.Timestamps.ReferenceAssemblyChangeRelative
  Expected: not equal to 0
  But was:  0
at Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeRelative() in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.TimeStamp.fs:line 262

```

dsyme commented:
Re things like 
```
3) Failed : Tests.LanguageService.TimeStamp.UsingMSBuild.Timestamps.ReferenceAssemblyChangeAbsolute
  Expected: not equal to 0
  But was:  0
```
Sometimes these are parse counts.  Improving the diagnostics from the tests may help

dsyme commented on src/fsharp/vs/service.fs:
```diff
@@ -1372,153 +1414,127 @@ module internal Parser =
+        let lower = String.lowercase fileName
+        let lightOnByDefault = List.exists (Filename.checkSuffix lower) FSharpLightSyntaxFileSuffixes
+        let lightSyntaxStatus = if lightOnByDefault then (options.LightSyntax <> Some false) else (options.LightSyntax = Some true)
+        LightSyntaxStatus(lightSyntaxStatus, true)
```
It feels like this code is duplicating ``ComputeLightSyntaxInitialStatus`` ?
 https://github.com/Microsoft/visualfsharp/blob/20573b7decc4989b625a769952c45a4c53fc510a/src/fsharp/CompileOps.fs#L3010

auduchinok commented:
@dsyme Do you have an idea what could be broken with these changes?
Most of the failures above are due to an absence of code completions for location and this looks like a problem with some caches.

I have reproduced no completion while editing an actual code and while debugging it I found that types of some values were lost in TAST, so no members were found for these values and that in turn lead to no completions. This behaviour could be reproduced with some particular namespaces opened only. I think I can try to shrink it to a project where it can be reproduced.

dsyme commented:
@auduchinok There are only 4 test failures.  I was wondering if they have something to do with ``#light`` at the top of a file since some of the source code involves that - not sure though.  Still looking around

dsyme commented:
@auduchinok I pushed a resolve commit to your branch

dsyme commented:
@auduchinok I've found the cause of some of the problems - it was to do with the changes to DependencyFiles - will push a commit soon

dsyme commented:
@auduchinok OK, fixes pushed, see https://github.com/Microsoft/visualfsharp/pull/3601/commits/8254407c3e87050eeae1018a894796285f16f25e

dsyme commented:
@KevinRansom Once this is green this is ready to go in

cartermp commented:
@Pilchie FYI this is one we should make a case to take with the remainder of @KevinRansom's work for Core SDK project support.  Big win for working in large files.

dsyme commented:
@cartermp No, I don't think we should make a special case for this to get it into VS at the end of a release phase - there is enough churn here that there is a slim chance of a regression of some kind - the perf gains are good but I'd count on letting it bake in master for a bit

dsyme commented:
Two more failures now! 
```
1) Error : Tests.LanguageService.AutoCompletion.UsingMSBuild.Filenames.MayBeDifferentlyCased
System.Exception : No Authoring Scope
   at Salsa.Salsa.Privates.SimpleOpenFile.DoIntellisenseRequest(BackgroundRequestReason parseReason) in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\Salsa.fs:line 1201
   at Salsa.Salsa.Privates.SimpleOpenFile.AutoCompleteAtCursorImpl(BackgroundRequestReason reason, FSharpOption`1 filterText) in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\Salsa.fs:line 1325
   at Salsa.Salsa.MSBuildTestFlavor.Salsa-Salsa-VsOps-AutoCompleteAtCursor(OpenFile file) in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\Salsa.fs:line 1540
   at Tests.LanguageService.AutoCompletion.UsingMSBuild.Filenames.MayBeDifferentlyCased() in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 4357

2) Error : Tests.LanguageService.AutoCompletion.UsingProjectSystem.Filenames.MayBeDifferentlyCased
System.Exception : No Authoring Scope
   at Salsa.Salsa.Privates.SimpleOpenFile.DoIntellisenseRequest(BackgroundRequestReason parseReason) in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\Salsa.fs:line 1201
   at Salsa.Salsa.Privates.SimpleOpenFile.AutoCompleteAtCursorImpl(BackgroundRequestReason reason, FSharpOption`1 filterText) in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\Salsa.fs:line 1325
   at Salsa.Salsa.MSBuildTestFlavor.Salsa-Salsa-VsOps-AutoCompleteAtCursor(OpenFile file) in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\Salsa.fs:line 1540
   at Tests.LanguageService.AutoCompletion.UsingMSBuild.Filenames.MayBeDifferentlyCased() in D:\j\w\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 4357
```

auduchinok commented:
@dsyme Thank you!
I'll look into this failure tomorrow.

dsyme commented:
I pushed two variations, both of which seem to fix that particular test, but one of which is more radical than the other

1. https://github.com/Microsoft/visualfsharp/pull/3601/commits/2f3298c026d7e085c0893cf3b8e5920b6f8eb3d8 removes the use of ``CheckFileInProjectAllowingStaleCachedResults`` in our old code (which is what these tests are running) in favour of ``CheckFileInProject``.  This may however cause other tests to fail

2. https://github.com/Microsoft/visualfsharp/pull/3601/commits/e0b2a134bed159814f7d5d6464dc004a5178dc70 is a more conservative fix that makes ``CheckFileInProjectAllowingStaleCachedResults`` create a builder - previously this old test code had assumed the builder was created by ``ParseFileInProject`` but now ``ParseFile`` doesn't necessarily do that

In either case I've marked ``CheckFileInProjectAllowingStaleCachedResults`` as deprecated, asking people to use ``CheckFileInProject`` instead

KevinRansom commented:
@dsyme, @cartermp  2017.5 Peview 2.0 is next and baked.  I think preview 3.0 this PR alongside the project reference stuff for dotnet core would be aggressive but useful, and give us some time to stabilize it in the product before the actual release.

Kevin

Krzysztof-Cieslak commented:
I'm happy to test it on Ionide users as soon as it gets merged and new version of FCS is released ;-)

dsyme commented:
@KevinRansom We should actually wait a bit before pulling.  The PR makes ``checker.ParseFile`` the fast entry point, and ``checker.ParseFileInProject`` becomes slower since the the FSharpParsingOptions are created repeatedly.  

``ParseFile`` is now used by Rider, ``ParseFileInProject`` is still used by Visual F# Tools (and presumably Ionide).  We should do the extra work to fully deprecate ``checker.ParseFileInProject`` and make sure the FSharpParsingOptions are created only once by the Visual F# Tools.  Ionide and FsAutoComplete can eventually update when they migrate nuget package

(Unlike other editing tools, Visual F# Tools always uses the latest FCS code, and so we can't integrate changes to FCS until we've updated VF# Tools to adjust for those changes)

KevinRansom commented:
Maybe I can look at what that would take then.

dsyme commented:
@KevinRansom I've updated this PR to use ParseFile instead of ParseFileInProject.  This means maintaining both a ``FSharpParsingOptions`` and ``FSharpProjectOptions`` in the table in ``LanguageService.fs`` - please take a look at https://github.com/Microsoft/visualfsharp/pull/3601/commits/0af0578ac3b055a5a4b3f42db1c7dea70460034f

The FSharpParsingOptions is easily computable from FSharpProjectOptions but we just need to make sure we amortize the cost of this and store it in that table.

dsyme commented:
@cartermp The latest additions should mean that  #3404 is fixed, see the removal of the comment here https://github.com/Microsoft/visualfsharp/pull/3601/files#diff-8fcecd80bdf6aa0e75cf6116b43b793bL32

dsyme commented:
> @cartermp The latest additions should mean that #3404 is fixed, see the removal of the comment here https://github.com/Microsoft/visualfsharp/pull/3601/files#diff-8fcecd80bdf6aa0e75cf6116b43b793bL32

And indeed I validated this by hand - I was able to set breakpoints immediately in every file even in the context of VisualFSharp.sln, with no delays at all

cartermp commented:
Brilliant.  I'll have to test this out again this evening.

dsyme commented:
@Krzysztof-Cieslak I've pushed  FCS 16.0.1 incorporating this feature (since it is very close to being incorporated into master) based off this tag and commit: https://github.com/Microsoft/visualfsharp/tree/FCS-16.0.1

cartermp commented:
🎉 

@brettfo is there a way to get a merge over into a 15.5 branch so that we can validate with the latest bits?  With nightlies still down (😢 ), this will be more difficult for folks to use, so I think that if we use it against the branch we're working with internally it'll get some decent experience-level validation.

brettfo commented:
@KevinRansom and I are planning on a merge from `master` to `vs2017-rtm` (our shipping branch) soon.  Since signing is still down we'll have to request a signed build, but then we can distribute the MSI/VSIX for testing.

dsyme commented:
@brettfo That would be great
